In [ ]:
import numpy as np

In [ ]:
np.__version__

'1.25.2'

Vamos recorrer ao mesmo array dos exercícios anteriores, que representa o desempenho de 5 atletas em 4 voltas de uma pista de corrida, em 3 dias diferentes.

In [ ]:
rng = np.random.default_rng(seed=1)

In [ ]:
array = rng.normal(loc=50, scale=3, size=60).reshape(3, 5, 4).round(3)

In [ ]:
array

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [41.867, 44.333, 49.476, 48.733],
        [50.641, 50.652, 56.354, 46.664],
        [48.867, 56.128, 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

In [ ]:
array.shape

(3, 5, 4)

👆 Observe pelo `shape` que são 3 dias, 5 atletas e 4 voltas.

#### `max` / `argmax`

Lembre-se que, como os valores representam o tempo, em segundos, que cada atleta levou para completar uma volta, quanto maior este valor, pior o desempenho do atleta.

Utilize `max` para encontrar o pior desempenho diário (eixo 0).

In [ ]:
array.max(axis=0)

array([[51.037, 52.465, 53.882, 53.02 ],
       [52.716, 51.339, 50.107, 51.743],
       [51.781, 52.674, 56.354, 51.64 ],
       [52.195, 56.128, 52.637, 51.989],
       [52.743, 49.94 , 50.502, 50.327]])

👆 O resultado é um array com 5 linhas (correspondente aos 5 atletas) e 4 colunas (correspondente às 4 voltas). Cada valor é o tempo máximo nos 3 dias, daquele atleta e daquela volta.

Repita o exercício com `argmax`.

In [ ]:
array.argmax(axis=0)

array([[0, 0, 1, 1],
       [0, 0, 2, 0],
       [2, 2, 1, 0],
       [2, 1, 2, 1],
       [2, 2, 1, 1]])

👆 Isto indica que, para o primeiro atleta e a primeira volta, (primeira linha e primeira coluna da visualização) o pior desempenho foi no dia 0; para o último atleta e a última volta, o pior desempenho foi no dia 1.

Agora utilize `max` e `argmax` para encontrar o pior desempenho por atleta (eixo 1), em cada dia e em cada volta.

In [ ]:
array.max(axis=1)

array([[52.716, 52.465, 50.991, 51.797],
       [50.641, 56.128, 56.354, 53.02 ],
       [52.743, 52.674, 52.637, 49.058]])

👆 Este é o pior desempenho dentre todos os atletas, nos 3 dias (linhas) e nas 4 voltas (colunas).

In [ ]:
array.argmax(axis=1)

array([[1, 0, 0, 3],
       [2, 3, 2, 0],
       [4, 2, 3, 4]])

👆 No primeiro dia (primeira linha), nas 4 voltas (colunas), os piores atletas foram os de índice 1, 0, 0 e 3, respectivamente.

Encontre também o pior desempenho geral.

In [ ]:
array.max()

56.354

#### `min`/`argmin`

Repita o exercício com os métodos `max`/`argmax` para encontrar os melhores desempenhos.

Interprete os resultados, de forma análoga à feita para os piores desempenhos.

In [ ]:
array.min(axis=0)

array([[46.318, 47.95 , 49.784, 46.091],
       [41.867, 44.333, 48.389, 48.481],
       [50.641, 50.652, 50.085, 46.664],
       [47.791, 48.496, 48.554, 46.785],
       [48.458, 45.056, 46.254, 49.058]])

In [ ]:
array.argmin(axis=0)

array([[2, 2, 2, 0],
       [1, 1, 0, 2],
       [1, 1, 0, 1],
       [0, 2, 0, 2],
       [1, 1, 2, 2]])

In [ ]:
array.min(axis=1)

array([[47.791, 49.123, 47.654, 46.091],
       [41.867, 44.333, 49.476, 46.664],
       [46.318, 47.95 , 46.254, 46.785]])

In [ ]:
array.argmin(axis=1)

array([[3, 4, 4, 0],
       [1, 1, 1, 2],
       [0, 0, 4, 3]])

In [ ]:
array.min()

41.867

#### `ptp`

Encontre a diferença entre o pior e o melhor desempenho, em cada dia (eixo 0).

In [ ]:
array.ptp(axis=0)

array([[ 4.719,  4.515,  4.098,  6.929],
       [10.849,  7.006,  1.718,  3.262],
       [ 1.14 ,  2.022,  6.269,  4.976],
       [ 4.404,  7.632,  4.083,  5.204],
       [ 4.285,  4.884,  4.248,  1.269]])

👆 No caso do atleta 0 (primeira linha), o desempenho dele flutuou 4.719 segundos na primeira volta, 4.515 segundos na segunda volta, e assim por diante.

Encontre a diferença entre o pior e o melhor desempenho, em cada volta (eixo 2).

In [ ]:
array.ptp(axis=2)

array([[6.374, 4.327, 1.555, 4.006, 2.465],
       [4.709, 7.609, 9.69 , 7.261, 5.446],
       [3.466, 1.805, 5.129, 5.852, 6.489]])

👆 No caso da volta de índice 1 (segunda coluna), a diferença de desempenho entre os atletas no primeiro dia foi de 4.327 segundo, 7.609 segundos no segundo dia e 1.805 segundos no terceiro.

#### `clip`

"Clipe" este array para que nenhum valor seja menor que 43 ou maior que 55. Exiba o array para visualizar as alterações.

In [ ]:
array_clip = array.clip(min=43, max=55)

In [ ]:
array_clip

array([[[51.037, 52.465, 50.991, 46.091],
        [52.716, 51.339, 48.389, 51.743],
        [51.094, 50.882, 50.085, 51.64 ],
        [47.791, 49.511, 48.554, 51.797],
        [50.119, 49.123, 47.654, 49.228]],

       [[50.024, 49.173, 53.882, 53.02 ],
        [43.   , 44.333, 49.476, 48.733],
        [50.641, 50.652, 55.   , 46.664],
        [48.867, 55.   , 51.94 , 51.989],
        [48.458, 45.056, 50.502, 50.327]],

       [[46.318, 47.95 , 49.784, 47.166],
        [49.705, 50.286, 50.107, 48.481],
        [51.781, 52.674, 50.963, 47.545],
        [52.195, 48.496, 52.637, 46.785],
        [52.743, 49.94 , 46.254, 49.058]]])

Confirme com `min` e `max` que a alteração foi feita conforme desejado.

In [ ]:
array_clip.min()

43.0

In [ ]:
array_clip.max()

55.0

#### `round`

Utilize `round` para arredondar os valores para 2 casas decimais.

In [ ]:
array.round(decimals=2)

array([[[51.04, 52.46, 50.99, 46.09],
        [52.72, 51.34, 48.39, 51.74],
        [51.09, 50.88, 50.08, 51.64],
        [47.79, 49.51, 48.55, 51.8 ],
        [50.12, 49.12, 47.65, 49.23]],

       [[50.02, 49.17, 53.88, 53.02],
        [41.87, 44.33, 49.48, 48.73],
        [50.64, 50.65, 56.35, 46.66],
        [48.87, 56.13, 51.94, 51.99],
        [48.46, 45.06, 50.5 , 50.33]],

       [[46.32, 47.95, 49.78, 47.17],
        [49.7 , 50.29, 50.11, 48.48],
        [51.78, 52.67, 50.96, 47.54],
        [52.2 , 48.5 , 52.64, 46.78],
        [52.74, 49.94, 46.25, 49.06]]])

Verifique também o arredondamento para 1 casa decimal.

In [ ]:
array.round(decimals=1)

array([[[51. , 52.5, 51. , 46.1],
        [52.7, 51.3, 48.4, 51.7],
        [51.1, 50.9, 50.1, 51.6],
        [47.8, 49.5, 48.6, 51.8],
        [50.1, 49.1, 47.7, 49.2]],

       [[50. , 49.2, 53.9, 53. ],
        [41.9, 44.3, 49.5, 48.7],
        [50.6, 50.7, 56.4, 46.7],
        [48.9, 56.1, 51.9, 52. ],
        [48.5, 45.1, 50.5, 50.3]],

       [[46.3, 48. , 49.8, 47.2],
        [49.7, 50.3, 50.1, 48.5],
        [51.8, 52.7, 51. , 47.5],
        [52.2, 48.5, 52.6, 46.8],
        [52.7, 49.9, 46.3, 49.1]]])

#### `sum`

Utilize `sum` para visualizar o tempo total gasto por cada atleta em cada prova, nos 3 dias. Ou seja, você deve somar o tempo no eixo 2, correspondente às voltas.


In [ ]:
array_sum_axis_2 = array.sum(axis=2)

In [ ]:
array_sum_axis_2

array([[200.584, 204.187, 203.701, 197.653, 196.124],
       [206.099, 184.409, 204.311, 208.924, 194.343],
       [191.218, 198.579, 202.963, 200.113, 197.995]])

👆 O array representa os 3 dias (3 linhas) e os 5 atletas (5 colunas).

Agora, a partir deste array, some os desempenhos dos 5 atletas ao longo dos 3 dias (eixo 0).

In [ ]:
array_sum_axis_2_axis_0 = array_sum_axis_2.sum(axis=0)

In [ ]:
array_sum_axis_2_axis_0

array([597.901, 587.175, 610.975, 606.69 , 588.462])

👆 O array representa a soma geral do desempenho dos 5 atletas.

#### `cumsum`

Utilize `cumsum` para encontrar o tempo acumulado de cada atleta, em cada dia, ao final de cada volta (eixo 2).

In [ ]:
array.cumsum(axis=2)

array([[[ 51.037, 103.502, 154.493, 200.584],
        [ 52.716, 104.055, 152.444, 204.187],
        [ 51.094, 101.976, 152.061, 203.701],
        [ 47.791,  97.302, 145.856, 197.653],
        [ 50.119,  99.242, 146.896, 196.124]],

       [[ 50.024,  99.197, 153.079, 206.099],
        [ 41.867,  86.2  , 135.676, 184.409],
        [ 50.641, 101.293, 157.647, 204.311],
        [ 48.867, 104.995, 156.935, 208.924],
        [ 48.458,  93.514, 144.016, 194.343]],

       [[ 46.318,  94.268, 144.052, 191.218],
        [ 49.705,  99.991, 150.098, 198.579],
        [ 51.781, 104.455, 155.418, 202.963],
        [ 52.195, 100.691, 153.328, 200.113],
        [ 52.743, 102.683, 148.937, 197.995]]])

👆 No segundo dia, o terceiro atleta (linha 2) acumulou tempos de 50.641, 101.293, 157.647 e 204.311.

#### `mean` / `var` / `std`

Encontre o desempenho médio por volta (eixo 2) para cada atleta, nos 3 dias.

In [ ]:
array.mean(axis=2)

array([[50.146  , 51.04675, 50.92525, 49.41325, 49.031  ],
       [51.52475, 46.10225, 51.07775, 52.231  , 48.58575],
       [47.8045 , 49.64475, 50.74075, 50.02825, 49.49875]])

👆 Desempenho médio por volta nos 3 dias (3 linhas), para os 5 atletas (5 colunas).

Exiba também a variância e o desvio padrão dos mesmos dados.

In [ ]:
array.var(axis=2)

array([[ 5.832173  ,  2.60505119,  0.31180869,  2.26545619,  0.7817865 ],
       [ 3.89384469,  9.84265569, 11.92300419,  6.6615875 ,  4.79485519],
       [ 1.63923875,  0.49570519,  3.77044719,  6.09171069,  5.36063069]])

In [ ]:
array.std(axis=2)

array([[2.41498923, 1.6140171 , 0.55839832, 1.50514324, 0.88418691],
       [1.97328272, 3.1373007 , 3.45297034, 2.58100513, 2.18971578],
       [1.28032759, 0.70406334, 1.94176394, 2.46813911, 2.31530358]])

Agora exiba as 3 medidas estatísticas agregadas por atleta (eixo 1).

In [ ]:
array.mean(axis=1)

array([[50.5514, 50.664 , 49.1346, 50.0998],
       [47.9714, 49.0684, 52.4308, 50.1466],
       [50.5484, 49.8692, 49.949 , 47.807 ]])

👆 Desempenho médio por atleta nos 3 dias (3 linhas), nas 4 voltas (4 colunas).

In [ ]:
array.var(axis=1)

array([[ 2.60449864,  1.490168  ,  1.48693464,  4.95666456],
       [ 9.92850184, 18.17600984,  6.03789536,  5.16203544],
       [ 5.53089424,  2.72292896,  4.3917508 ,  0.7086572 ]])

In [ ]:
array.std(axis=1)

array([[1.61384592, 1.22072438, 1.21939929, 2.22635679],
       [3.15095253, 4.26333318, 2.45721293, 2.27201132],
       [2.35178533, 1.65012998, 2.09565045, 0.8418178 ]])

#### `all` / `any`

Primeiro, gere uma máscara booleana (array com valores `True`/`False`) para verificar quais valores em `array` são maiores que 50 segundos.

In [ ]:
array_mask = array > 50

In [ ]:
array_mask

array([[[ True,  True,  True, False],
        [ True,  True, False,  True],
        [ True,  True,  True,  True],
        [False, False, False,  True],
        [ True, False, False, False]],

       [[ True, False,  True,  True],
        [False, False, False, False],
        [ True,  True,  True, False],
        [False,  True,  True,  True],
        [False, False,  True,  True]],

       [[False, False, False, False],
        [False,  True,  True, False],
        [ True,  True,  True, False],
        [ True, False,  True, False],
        [ True, False, False, False]]])

Agora, use `all` para verifcar se algum atleta levou mais que 50 segundos para completar alguma volta (eixo 2), em qualquer dia.

In [ ]:
array_mask.all(axis=2)

array([[False, False,  True, False, False],
       [False, False, False, False, False],
       [False, False, False, False, False]])

👆 Isto indica que, no primeiro dia (primeira linha), o atleta de índice 2 (terceira coluna) demorou mais que 50 segundos em todas as voltas.

Utilize `any` para repetir a análise, mas considerando o tempo gasto em qualquer volta.

In [ ]:
array_mask.any(axis=2)

array([[ True,  True,  True,  True,  True],
       [ True, False,  True,  True,  True],
       [False,  True,  True,  True,  True]])

👆 No primeiro dia, todos os atletas tiveram pelo menos uma volta que levou mais que 50 segundos; no segundo dia, só o atleta de índice 1 não teve um desempenho pior que 50 segundos.